middleware/logger.go

```diff
@@ -13,22 +13,43 @@ import (
 	"time"
 )
 
+// DefaultLogFormatter contains all the data needed to construct a log message
+type DefaultLogFormatter struct {
+	buf bytes.Buffer
+}
+
+// LogFormatter allows you to customize this middleware's output format.
+type LogFormatter interface {
+	FormatRequest(r *http.Request) LogFormatter
+	FormatResponse(status, bytes int, elapsed time.Duration)
+	Log()
+	Recover(err error)
```

Comment:
You're right, this could be simplified down to one method, like:

    FormatLog(r *http.Request, status, bytes int, elapsed time.Duration, err error)

With `err == nil` meaning "log normal req/res info", otherwise ignore `status, bytes, elapsed` and show the error and stack trace.